Nodetool RPC handling fixes

swvist opened the pull request.

swvist commented:
Some of the changes here are also applicable to `rpcterms` which I haven't done. I need to figure out what it actually does.

tsloughter commented on priv/templates/nodetool:
```diff
@@ -32,17 +32,10 @@ main(Args) ->
             io:format("~p\n", [rpc:call(TargetNode, init, restart, [], 60000)]);
         ["reboot"] ->
             io:format("~p\n", [rpc:call(TargetNode, init, reboot, [], 60000)]);
+        ["rpc", Module, Function | RpcArgs] when RpcArgs == [] ->
```
this can just be `["rpc", Module, Function]` since `[]` would mean simply the end of the list.

swvist commented:
@tsloughter ah yes! Fixed them. Also, do you have an example 'rpcterms` usage?

tsloughter commented:
@swvist `rpcterms` is actually used in the start script, `relx_nodetool rpcterms net_kernel get_net_ticktime`. It takes an erlang term as a string which it parses it into a list the same as you'd expect from consulting a file of terms.

swvist commented:
@tsloughter, I have refactored `rpcterms` to be similar to `rpc` and added a test suite for `nodetool`.

I am not too happy with this though. I think `rpc` and `rpcterms` are too similar in intent and both of them suffer from limitations that the other form cannot help with. 

Currently `rpc` can run functions with zero, one or many arguments, with the limitation of arguments being strings. `rpcterms` can run functions with zero or one argument, but the argument can be any Erlang term. It is not possible to run an Erlang function that takes multiple arguments and can be any valid Erlang term. (See 5 below).

I propose the following changes:

1) Make `rpc` behave like `rpcterms` and drop `rpcterms`.
2) All `rpc` will be of apply(M, F, ArgsList) form.

After the changes, the following should work:

1.  `./release-name rpc erlang node`
2. `./release-name rpc atom_to_list "[atom]"`
3. `./release-name rpc erlang tuple_size "[{abc, def}]"`
4. `./release-name rpc string concat "[\"abc\", "\def"\]"`
5. `./release-name rpc erlang apply "[erlang, apply, [lists, append, [[1, 2], [3, 4]]]]"`

djnym commented:
FWIW, I also added 'eval' which allows a more natural way of doing the same thing, so for your examples you could do things like

1. ./release-name eval 'erlang:node()'
2. ./release-name eval 'atom_to_list(atom)'
3. ./release-name eval 'erlang:tuple_size({abc,def})'
4. ./release-name eval 'string:concat (["abc","def"])'
5. ./release-name eval 'erlang:apply(erlang, apply, [lists, append, [[1,2],[3,4]]])'

I find this form more natural as I don't have to separate out all the arguments, and can just have a single level of escaping in the shell.  That doesn't meant that rpc or rpcterms shouldn't exist, but may not need too?

lrascao commented:
@swvist Could you please rebase onto master?
